File: tests/refs/dijkstra_shortest_path.py
"""Reference: Dijkstra shortest paths from `src` on a non-negative weighted graph.

Input edges are an undirected list `[u, v, w]`. Returns a list of integer
distances of length V; unreachable nodes report -1 (instead of a float `inf`,
to keep JSON-serialisable fixtures).
"""

from __future__ import annotations
import heapq
import math
from typing import List, Tuple


def dijkstra(V: int, edges: List[Tuple[int, int, int]], src: int) -> List[int]:
    adj: List[List[Tuple[int, int]]] = [[] for _ in range(V)]
    for u, v, w in edges:
        adj[u].append((v, w))
        adj[v].append((u, w))

    INF = math.inf
    dist: List[float] = [INF] * V
    dist[src] = 0
    pq: List[Tuple[float, int]] = [(0, src)]
    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]:
            continue
        for v, w in adj[u]:
            nd = d + w
            if nd < dist[v]:
                dist[v] = nd
                heapq.heappush(pq, (nd, v))

    return [-1 if d == INF else int(d) for d in dist]

File: tests/refs/DijkstraShortestPath.java
/*
 * Reference Java implementation for tests/cases/dijkstra_shortest_path.json.
 *
 * Dijkstra on an undirected non-negative weighted graph. Unreachable nodes
 * are reported as -1 to keep results JSON-friendly.
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;

public class DijkstraShortestPath {
    public static List<Long> dijkstra(long V, long[][] edges, long src) {
        int n = (int) V;
        List<long[]>[] adj = new List[n];
        for (int i = 0; i < n; ++i) adj[i] = new ArrayList<>();
        for (long[] e : edges) {
            int u = (int) e[0], v = (int) e[1];
            long w = e[2];
            adj[u].add(new long[]{v, w});
            adj[v].add(new long[]{u, w});
        }
        long INF = Long.MAX_VALUE;
        long[] dist = new long[n];
        Arrays.fill(dist, INF);
        dist[(int) src] = 0;
        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        pq.offer(new long[]{0, src});
        while (!pq.isEmpty()) {
            long[] cur = pq.poll();
            long d = cur[0];
            int u = (int) cur[1];
            if (d > dist[u]) continue;
            for (long[] nb : adj[u]) {
                int v = (int) nb[0];
                long nd = d + nb[1];
                if (nd < dist[v]) {
                    dist[v] = nd;
                    pq.offer(new long[]{nd, v});
                }
            }
        }
        List<Long> out = new ArrayList<>(n);
        for (long d : dist) out.add(d == INF ? -1L : d);
        return out;
    }
}
